metadata/src/main/java/org/apache/kafka/timeline/BaseHashTable.java

```diff
@@ -56,12 +61,27 @@
         this.elements = new Object[expectedSizeToCapacity(expectedSize)];
     }
 
+    /**
+     * Calculate the capacity we should provision, given the expected size.
+     *
+     * Our capacity must always be a power of 2, and never less than 2.
+     */
     static int expectedSizeToCapacity(int expectedSize) {
-        if (expectedSize <= 1) {
-            return 2;
+        if (expectedSize >= MAX_CAPACITY / 2) {
+            return MAX_CAPACITY;
+        }
+        return Math.max(MIN_CAPACITY, roundUpToPowerOfTwo(expectedSize * 2));
+    }
+
+    private static int roundUpToPowerOfTwo(int i) {
+        if (i <= 0) {
+            return 0;
+        } else if (i > MAX_SIGNED_POWER_OF_TWO) {
```

Comment:
Are all negative numbers acceptable? If so, the negative numbers which get positive by overflow (like `-1610612735`) encounter error.

Reply:
This function just returns 0 if the number is less than or equal to 0, so we don't have to consider that case.

Reply:
Sorry for my unclear comment.

I assumed that all negative numbers are acceptable for `expectedSizeToCapacity` because of test case `assertEquals(2, BaseHashTable.expectedSizeToCapacity(-123));`. However, some magic numbers (for example, `-1610612735`) can produce `ArithmeticException`.

At any rate, above case are nonexistent in latest commit now.

Reply:
Good catch.  Yes, this should be fixed now in the latest logic.